gemini/use-cases/retrieval-augmented-generation/qna_using_query_routing/utils/qna_using_query_routing_utils.py

```diff
@@ -0,0 +1,163 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility file with more generic fuctions"""
+
+import configparser
+from datetime import datetime
+
+# Utils
+import os
+from typing import List
+
+from google.cloud import aiplatform
+from langchain.document_loaders import TextLoader, UnstructuredPDFLoader
+from langchain.text_splitter import CharacterTextSplitter
+from langchain_core.documents.base import Document
+
+config_file = "config.ini"
+config = configparser.ConfigParser()
+config.read(config_file)
+
+
+def get_deployed_index_id(
+    me_index_name: str, location: str
+) -> tuple[aiplatform.MatchingEngineIndexEndpoint, str]:
+    """
+    Retrieves the deployed index ID and index endpoint from Vector Search.
+
+    Checks if a vector search index with the specified name is already deployed in the given location.
+
+    Args:
+        me_index_name (str): The name of the Vector Search index.
+        location (str): The location where the index is deployed.
+
+    Returns:
+        tuple: A tuple containing:
+            * aiplatform.MatchingEngineIndexEndpoint: The index endpoint object, or None if not found.
+            * str: The deployed index ID, or None if not found.
+    """
+
+    index_endpoint_name = me_index_name + "-endpoint"
+
+    list_endpoints = aiplatform.MatchingEngineIndexEndpoint.list(
+        filter=f"display_name={index_endpoint_name}", location=location
+    )
+    if list_endpoints:
+        index_endpoint_resource_name = list_endpoints[0].resource_name
+
+        index_endpoint = aiplatform.MatchingEngineIndexEndpoint(
+            index_endpoint_name=index_endpoint_resource_name
+        )
+        if index_endpoint.deployed_indexes:
+            deployed_index_id = index_endpoint.deployed_indexes[0].id
+        else:
+            deployed_index_id = None
+            print(
+                f"Index endpoint resource is not available for: {index_endpoint_resource_name}"
+            )
+    else:
+        index_endpoint = deployed_index_id = None
+        print(f"Index endpoint is not available for: {index_endpoint_name}")
+
+    return index_endpoint, deployed_index_id
+
+
+def get_split_documents(index_path: str) -> List[Document]:
+    """
+    Loads documents from a folder and splits them into manageable chunks.
+
+    Supports both PDF and plain text documents.
+
+    Args:
+        index_path (str): The path to the folder containing the document(s).
+
+    Returns:
+        List['str']: A list of strings, representing the chunked document(s).
+    """
+
+    split_docs = []
+
+    if index_path[-1] != "/":
+        index_path = index_path + "/"
+    if index_path == "":
+        index_path = "."
```

Comment:
I'd recommend using the pathlib or something similar to handle these path operations.